app/models/tweet_fetcher.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: tweet_fetchers
+#
+#  id          :bigint           not null, primary key
+#  backoff     :integer          default(1)
+#  complete    :boolean          default(FALSE)
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#  data_set_id :bigint
+#  user_id     :string
+#
+# Indexes
+#
+#  index_tweet_fetchers_on_data_set_id  (data_set_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (data_set_id => data_sets.id)
+#
+
+# TweetFetcher fetches tweets for a single user. One DataSet spawns many
+# TweetFetchers, one per user_id in the cohort. This allows us to execute
+# tweet fetching out-of-band, so that we can be sensitive to rate limits.
```

Comment:
One TweetFetcher per user doesn't seem right to me. It ties cohort data collection too closely to DataSet. I think I'd prefer a DataSet to have a single TweetFetcher (later to become a subclass of DataFetcher?). The TweetFetcher would then be handed all the user ids, and it would figure out what to do with them. The reason I like this more is because it would allow us to do things like create a TweetStreamFetcher which collects tweets using a streaming API query instead of retrospective timeline fetching, or create a CrimsonHexagonFetcher which uses Crimson Hexagon as a source instead of Twitter. Both of those collection methods are capable of fetching many users in a single query, so constraining the API to a single user id seems too limiting.

Right now, because we are only using a single API key and the bottleneck is throttling, having multiple fetchers at once doesn't give us a speedup. If we did want to parallelize this process with, say, multiple API keys, I'd prefer for the management of that parallelization (e.g. distribution of API keys to various workers) to be in a single TweetFetcher rather than baked into how TweetFetchers are created. Having a single API distribution mechanism with multiple keys behind it is how I'm doing it in Python and the API is clean as heck, but don't tell. 😉

Reply:
Hm, so you're saying the TweetFetcher should be more of a factory that can then delegate to a variety of fetching mechanisms, without altering the API that DataSet uses? OK, I think I'm persuaded here.

I'm using a bunch of them right now because (since I can only fetch one timeline at a time) that lets me queue them up for rate-limiting purposes. But I could transfer that rate-limiting functionality to the jobs themselves, and have the TweetFetcher spawn all the jobs.

OK, yes, this makes sense. Thanks for the idea.